File: java/src/main/java/com/example/doubleup/service/WithdrawService.java
package com.example.doubleup.service;

import com.example.doubleup.dao.PlayerRepository;
import com.example.doubleup.dao.WithdrawalRepository;
import com.example.doubleup.enums.AuditOperation;
import com.example.doubleup.model.Player;
import com.example.doubleup.model.Withdrawal;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.OffsetDateTime;

@Service
public class WithdrawService {

    @Autowired
    private PlayerRepository playerRepository;

    @Autowired
    private WithdrawalRepository withdrawalRepository;

    @Autowired
    private AuditLogService auditLogService;

    @Transactional
    public String withdraw(Player player) {
        Withdrawal withdrawal = new Withdrawal(player, OffsetDateTime.now(), player.getMoneyInPlay());
        withdrawal = withdrawalRepository.save(withdrawal);
        auditLogService.writeAuditLog(player, AuditOperation.INSERT, withdrawal.getId(), "withdrawal");

        player.setAccountBalance(player.getAccountBalance() + player.getMoneyInPlay());
        player.setMoneyInPlay(0L);
        player = playerRepository.save(player);
        auditLogService.writeAuditLog(player, AuditOperation.UPDATE, player.getId(), "player");

        return "OK";
    }
}

File: go/service/withdraw_service.go
package service

import (
    "fmt"
    "database/sql"
    "double_up/dao"
    "double_up/enums"
    "double_up/model"
)

func Withdraw(db *sql.DB, player *model.Player) (string, error) {
    withdrawal := model.NewWithdrawal(player.ID, player.MoneyInPlay)
    err := dao.CreateWithdrawal(db, withdrawal)
    if err != nil {
        fmt.Println(err)
        return "ERROR", err
    }
    WriteAuditLog(db, player, enums.Insert, withdrawal.ID, "withdrawal")

    player.AccountBalance += player.MoneyInPlay
    player.MoneyInPlay = 0
    _, err = dao.UpdatePlayer(db, player)
    if err != nil {
        fmt.Println(err)
        return "ERROR", err
    }
    WriteAuditLog(db, player, enums.Update, player.ID, "player")

    return "OK", nil
}
